subsys/bluetooth/host/settings.c

```
@@ -33,75 +33,127 @@
+int bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+			   const bt_addr_le_t *addr, const char *key)
 {
 	size_t len = 3;
 
-	/* Skip if path_size is less than 3; strlen("bt/") */
+	/* path_size is less than 3; strlen("bt/") */
+	BT_ASSERT(path_size >= len);
+
+	/* Key format:
+	 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
+	 */
+	strcpy(path, "bt/");
+
+	/* Concatenate subsys as much the free space permits */
+	strncpy(&path[len], subsys, (path_size - len));
+
+	/* Postfix '/' if there is free space */
+	len = strlen(path);
 	if (len < path_size) {
-		/* Key format:
-		 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
+		path[len] = '/';
+		len++;
+	}
+
+	/* Concatenate addr as much the free space permits */
+	for (int8_t i = 5; i >= 0 && len < path_size; i--) {
+		len += bin2hex(&addr->a.val[i], 1, &path[len],
+			       path_size - len);
+	}
+
+	/* Postfix addr type if there is free space */
+	if (len < path_size) {
+		/* Type can be either BT_ADDR_LE_PUBLIC or
+		 * BT_ADDR_LE_RANDOM (value 0 or 1)
 		 */
-		strcpy(path, "bt/");
-		strncpy(&path[len], subsys, path_size - len);
-		len = strlen(path);
-		if (len < path_size) {
-			path[len] = '/';
-			len++;
-		}
+		path[len] = '0' + addr->type;
+		len++;
+	}
 
-		for (int8_t i = 5; i >= 0 && len < path_size; i--) {
-			len += bin2hex(&addr->a.val[i], 1, &path[len],
-				       path_size - len);
-		}
+	/* Postfix '/' and concatenate key as much the free space permits */
+	if (key && len < path_size) {
+		path[len] = '/';
+		len++;
+		strncpy(&path[len], key, path_size - len);
+		len += strlen(&path[len]);
+	}
 
-		if (len < path_size) {
-			/* Type can be either BT_ADDR_LE_PUBLIC or
-			 * BT_ADDR_LE_RANDOM (value 0 or 1)
-			 */
-			path[len] = '0' + addr->type;
-			len++;
-		}
+	/* If path string is full, always null terminate at path_size */
+	if (len >= path_size) {
+		/* Truncate string */
+		path[path_size - 1] = '\0';
+	}
```

There is where I think we need to do something different. If the buffer isn't large enough to hold the content, we shouldn't just silently truncate it

---

Should I replace with `BT_ASSERT(len < path_size)` ?

---

Something like that. I really think that we shouldn't silently truncate data like this. Not sure if a runtime error (returning an error to the settings subsystem) or an assert is best